src/components/EnhancedMap/ImageMarkerLayer/ImageMarkerLayer.jsx

```diff
@@ -50,85 +64,129 @@ const ImageMarkerLayer = props => {
   buildIcon: PropTypes.func,
   markerColor: PropTypes.string,
 }
 
-class MapillaryViewer extends Component {
-  containerRef = createRef();
+const MapillaryViewer = ({ initialImageKey }) => {
+  const containerRef = useRef()
 
-  componentDidMount() {
-    this.viewer = new Viewer({
-      accessToken: getAccessToken(),
-      container: this.containerRef.current,
-      imageId: this.props.initialImageKey,
-      component: { cover: false },
-    })
-  }
+  useEffect(() => {
+    if (!initialImageKey) {
+      console.error('Initial image key is null or undefined')
+      return
+    }
+
+    let viewer
+    try {
+      const accessToken = getAccessToken();
+      if (!accessToken) {
+        console.error('Access token is not available');
+        return;
+      }
 
-  componentWillUnmount() {
-    if (this.viewer) {
-      this.viewer.remove();
+      if (imageCache.has(initialImageKey)) {
+        viewer = imageCache.get(initialImageKey);
+      } else {
+        viewer = new Viewer({
+          accessToken: accessToken,
+          container: containerRef.current,
+          imageId: initialImageKey,
+        });
+        imageCache.set(initialImageKey, viewer);
+      }
+    } catch (error) {
+      console.error('Error initializing Mapillary viewer:', error)
     }
-  }
 
-  shouldComponentUpdate(nextProps) {
-    return nextProps.initialImageKey !== this.props.initialImageKey
-  }
+    return () => {
+      try {
+        if (viewer) {
+          viewer.remove()
+        }
+      } catch (error) {
+        console.error('Error removing Mapillary viewer:', error)
+      }
+    }
+  }, [initialImageKey])
 
-  render() {
-    return (
-      <div className="mr-p-2 mr-pt-4 mr-relative">
-        <div ref={this.containerRef} id="mapillary-viewer" style={{ width: 335, height: 263 }}></div>
-      </div>
-    )
-  }
+  return (
+    <div className="mr-p-2 mr-pt-4 mr-relative">
+      <div ref={containerRef} id="mapillary-viewer" className="mr-w-full mr-h-64"></div>
+    </div>
+  )
 }
 
-const buildImageMarkers = (images, icon, imageClicked, imageAlt, layerId, layerLabel) => {
-  if (!images || images.length === 0) {
+const buildImageMarkers = (images, icon, markerColor, imageClicked, layerId, layerLabel) => {
+  try {
+    if (!images || images.length === 0) {
+      return []
+    }
+
+    return _map(images, imageInfo => {
+      if (!imageInfo || (!imageInfo.position?.lat || !imageInfo.position?.lon) && (!imageInfo.lat || !imageInfo.lon)) {
+        console.error(`Invalid position for image key: ${imageInfo?.key}`, imageInfo)
+        return null
+      }
+
+      if (!imageInfo.url) {
+        console.error(`Invalid URL for image key: ${imageInfo.key}`, imageInfo)
+        return null
+      }
+
+      return (
+        <Marker
+          key={imageInfo.key}
+          mrLayerId={layerId}
+          mrLayerLabel={layerLabel}
+          position={[imageInfo.position?.lat || imageInfo.lat, imageInfo.position?.lon || imageInfo.lon]}
```

Comment:
This line has a similar bug to what I described above; if `imageInfo.position.lat == 0` then `imageInfo.lat` (which is probably `undefined`) will be used instead. If you define a `getLatLng` helper you could reuse it here to simplify and correct the code.